src/ttvchat/TwitchChatService.ts

```diff
@@ -0,0 +1,130 @@
+
+  private log: log;
+  private loginStatusBarItem: vscode.StatusBarItem;
+  private chatClientStatusBarItem: vscode.StatusBarItem;
+  private chatClient: ChatClient;
+  private config?: vscode.WorkspaceConfiguration;
+
+  constructor(api: HighlighterAPI, outputChannel: vscode.OutputChannel) {
+    this.log = new Logger(outputChannel).log;
+    this.chatClient = new ChatClient(this.log);
+    this.chatClient.disconnect.bind(this.chatClient);
+
+    this.config = vscode.workspace.getConfiguration(Configuration.sectionIdentifier);
+
+    this._api = api;
+    this._authenticationService = new AuthenticationService(this.log);
+    this._authenticationService.signOutHandler.bind(this._authenticationService);
+
+    this.loginStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
+    this.loginStatusBarItem.text = `$(sign-in) twitch`;
+    this.loginStatusBarItem.command = Commands.signIn;
+    this.loginStatusBarItem.tooltip = 'Twitch Line Highlighter Login';
+
+    this.chatClientStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
+    this.chatClientStatusBarItem.text = `$(plug) Disconnected`;
+    this.chatClientStatusBarItem.command = Commands.connect;
+    this.chatClientStatusBarItem.tooltip = 'Twitch Line Highlighter Chat Bot';
+  }
+
+  public async initialize(context: vscode.ExtensionContext): Promise<void> {
+    this.log('ttvchat initializing...');
+
+    context.subscriptions.push(
+      this.loginStatusBarItem,
+      this.chatClientStatusBarItem,
+
+      vscode.workspace.onDidChangeConfiguration(this.onDidChangeConfigurationHandler, this),
+
+      this._authenticationService.onAuthStatusChanged(this.onAuthStatusChangedHandler, this),
+
+      this.chatClient.onChatClientConnected(this.onChatClientConnectedHandler, this),
+      this.chatClient.onChatClientMessageReceived(this.onChatClientMessageReceivedHandler, this),
+
+      vscode.commands.registerCommand(Commands.signIn, this._authenticationService.signInHandler, this._authenticationService),
+      vscode.commands.registerCommand(Commands.signOut, this.onSignOutHandler, this),
+
+      vscode.commands.registerCommand(Commands.connect, this.chatClient.connect, this.chatClient),
+      vscode.commands.registerCommand(Commands.disconnect, this.chatClient.disconnect, this.chatClient)
+    );
+
+
+    this.chatClient.initialize(context);
+    await this._authenticationService.initialize();
+
+    this.log('ttvchat initialized.');
+  }
+
+  private onDidChangeConfigurationHandler(event: vscode.ConfigurationChangeEvent) {
+    if (!event.affectsConfiguration(Configuration.sectionIdentifier)) {
+      return;
+    }
+    this.config = vscode.workspace.getConfiguration(Configuration.sectionIdentifier);
+  }
+
+  private onAuthStatusChangedHandler(signedIn: boolean) {
+    if (signedIn) {
+      vscode.window.showInformationMessage('Twitch token has been saved.');
```

Comment:
This seems to show up every time I load up VS Code and the extension is enabled. Maybe we could set things up to only show this on first time signing in?

Reply:
I'll look into the possibility of moving the notification to the block of code where the token is actually saved and then only write an entry into the output window here.

Reply:
I ended up just logging that the token was saved.

As an aside, the reason it saves when you open a new instance of VS CODE is that it verifies the token each time (and every hour thereafter), and when the verification returns it includes a new token.

Reply:
Looks good now. Thanks!